cdap-common/src/main/java/org/apache/twill/internal/yarn/Hadoop21YarnAppClient.java

```diff
@@ -50,21 +54,44 @@
  * </p>
  */
 // TODO: This is a copy from Twill, and can be removed when TWILL-119 is fixed. CDAP-4923 tracks Twill-0.8.0 upgrade.
+// After copying, we have removed the addRMToken() method. That functionality is in YarnTokenUtils.
+// Additionally, we have updated it to not have a single YarnClient, but instead a collection of YarnClients, one per
+// UGI. We need to do this because YarnClient is not designed to be reusable across UGIs. Even if the calling UGI
+// changes, the Yarn user still is the original UGI that constructed the YARN client.
 public final class Hadoop21YarnAppClient extends AbstractIdleService implements YarnAppClient {
 
   private static final Logger LOG = LoggerFactory.getLogger(Hadoop21YarnAppClient.class);
-  private final YarnClient yarnClient;
+  private final Configuration configuration;
+
+  // TODO: make thread safe
+  // alternatively, simply construct a new YARNClient when needed. Shouldn't be too expensive
+  private Map<String, YarnClient> yarnClients = new HashMap<>();
 
   public Hadoop21YarnAppClient(Configuration configuration) {
-    this.yarnClient = YarnClient.createYarnClient();
-    yarnClient.init(configuration);
+    this.configuration = configuration;
+  }
+
+  private YarnClient getYarnClient() {
+    try {
+      String userName = UserGroupInformation.getCurrentUser().getShortUserName();
+      if (!yarnClients.containsKey(userName)) {
+        YarnClient yarnClient = YarnClient.createYarnClient();
+        yarnClient.init(configuration);
+        yarnClient.start();
+
+        yarnClients.put(userName, yarnClient);
+      }
+      return yarnClients.get(userName);
+    } catch (IOException e) {
```

Comment:
true, but then I need to declare the variable outside the try block, so I can un-ident at that cost